### PR TITLE
ghost nodes for cylc graph.

### LINKED ISSUE #2287: graph implicit dependence as ghost nodes
Ref: https://github.com/cylc/cylc/issues/2280#issuecomment-301932374

### PR DESCRIPTION
**Pull request by oliver-sanders**
Closes #2287 

Graph view now displays tasks using ovals rather than boxes to match gcylc.

__Before__
before

__After__
after
ng)

### CODE REVIEW

File: lib/cylc/cylc_xdot.py [hunk 1]
```diff
@@ -401,9 +402,18 @@ def get_graph(self, group_nodes=[], ungroup_nodes=[],
...
-            name = TaskID.split(node.get_name())[0]
+            name, point = TaskID.split(node.get_name())
             if name in family_nodes:
+                # Style family nodes.
                 node.attr['shape'] = 'doubleoctagon'
+            else:
+                # Style regular (task) nodes.
+                node.attr['shape'] = 'oval'
+            if not any(seq.is_on_sequence(get_point(point)) for seq in
+                       self.suiterc.taskdefs[name].sequences):
```
Comment (matthewrmshin) [inline lib/cylc/cylc_xdot.py]: In essence of our recent validate optimisation, we know that sequence calculations are expensive. I wonder if it is worth caching the results here?

File: lib/cylc/cylc_xdot.py [hunk 2]
```diff
@@ -382,11 +390,27 @@ def group_all(self, w):
...
+        """Returns True if the task <name> at cycle point <point> is a ghost.
+        """
+        for sequence in self.suiterc.taskdefs[name].sequences:
+            p_str = str(point)
+            if (cache and sequence in cache and p_str in cache[sequence] and
+                    cache[sequence][p_str]):
+                return False
+            else:
+                temp = sequence.is_on_sequence(get_point(point))
+                if cache:
```
Comment (matthewrmshin) [inline lib/cylc/cylc_xdot.py]: This won't work if `cache` is empty?

**Comment by matthewrmshin** [comment]
(Branch in conflict.)

**Comment by matthewrmshin** [comment]
(`./tests/graphing/09-close-fam.t` failing on Travis CI.)

**Comment by hjoliver** [comment]
(conflicts)

**Comment by oliver-sanders** [comment]
(resolutions)

**Comment by hjoliver** [comment]
(heh heh :grin: )

File: tests/graphing/03-filter/graph.plain.ref.orig [hunk 3]
```diff
@@ -21,22 +21,22 @@ edge "elderberry.1" "fava.1" solid
...
-node "cherry.1" "cherry\n1" unfilled box black
-node "coconut.1" "coconut\n1" unfilled box black
-node "date.1" "date\n1" unfilled box black
-node "durian.1" "durian\n1" unfilled box black
-node "elderberry.1" "elderberry\n1" unfilled box black
-node "fava.1" "fava\n1" unfilled box black
-node "fennel.1" "fennel\n1" unfilled box black
-node "grape.1" "grape\n1" unfilled box black
+node "artichoke.1" "artichoke\n1" unfilled ellipse black
+node "arugula.1" "arugula\n1" unfilled ellipse black
```
Comment (hjoliver) [inline tests/graphing/03-filter/graph.plain.ref.orig]: (box/ellipse ... and at first I was slightly concerned about the +/- 900 lines change size)

**Comment by hjoliver** [comment]
I thought you had broken right-click group/ungroup of families, but it turns out I'd checked out your other branch with identical branch name minus the final 'w' in '-view'!   1 test failing, but otherwise looks good to me (tested with nested families too).

File: tests/graphing/03-filter/graph.plain.ref.orig (hunk 3 above)
Comment (oliver-sanders) [inline tests/graphing/03-filter/graph.plain.ref.orig]: Yes, its been fun changing all those lines!

**Comment by hjoliver** [comment]
Can be merged if tests pass now.  Over 'n out.

**Comment by hjoliver** [comment]
Actually, you could mention this at the end of the CUG section "How The Graph Determines Task Instantiation".

**Comment by matthewrmshin** [comment]
I'll take the liberty to merge this.

**Pull request merged**